design/src/rfcs/rfc0027_improve_sdk_request_id_access.md

```diff
@@ -0,0 +1,244 @@
+This logic will try different request ID header names in order of probability
+since AWS services have a couple of header name variations. `x-amzn-requestid` is
+the most common, with `x-amzn-request-id` being the second most common.
+
+`aws-http` will also implement `RequestId` for `aws_smithy_types::error::Error`,
+and the `request_id` method will be removed from `aws_smithy_types::error::Error`.
+Places that construct `Error` will place the request ID into its `extras` field,
+where the `RequestId` trait implementation can retrieve it.
+
+A codegen decorator will be added to `sdk-codegen` to implement `RequestId` for
+operation errors, and the existing `request_id` accessors will be removed from
+`CombinedErrorGenerator` in `codegen-core`.
+
+With these changes, customers can directly access request IDs from `SdkError` and
+operations errors by importing the `RequestId` trait. Additionally, the Smithy/SDK
+purity is improved since both places where request IDs are leaked to Smithy clients
+will be resolved.
+
+### Implement `RequestId` for outputs
+
+To make it possible to retrieve request IDs when using the fluent client, the new
+`RequestId` trait can be implemented for outputs.
+
+Some services (e.g., Transcribe Streaming) model the request ID header in their
+outputs, while other services (e.g., Directory Service) model a request ID
+field on errors. In some cases, services take `RequestId` as a modeled input
+(e.g., IoT Event Data). It follows that it is possible, but unlikely, that
+a service could have a field named `RequestId` that is not the same concept
+in the future.
+
+Thus, name collisions are going to be a concern for putting a request ID accessor
+on output. However, if it is implemented as a trait, then this concern is partially
+resolved. In the vast majority of cases, importing `RequestId` will provide the
+accessor without any confusion. In cases where it is already modeled and is the
+same concept, customers will likely just use it and not even realize they didn't
+import the trait. The only concern is future cases where it is modeled as a
+separate concept, and as long as customers don't import `RequestId` for something
+else in the same file, that confusion can be avoided.
+
+In order to implement `RequestId` for outputs, either the original response needs
+to be stored on the output, or the request ID needs to be extracted earlier and
+stored on the output. The latter will lead to a small amount of header lookup
+code duplication.
+
+In either case, the `StructureGenerator` needs to be customized in `sdk-codegen`
+(Appendix B outlines an alternative approach to this and why it was dismissed).
+This will be done by adding customization hooks to `StructureGenerator` similar
+to the ones for `ServiceConfigGenerator` so that a `sdk-codegen` decorator can
+conditionally add fields and functions to any generated structs. A hook will
+also be needed to additional trait impl blocks.
```

Comment:
would be interested in the design here—I'd probably envision something like the following but I'm curious about your thoughts

```rust
struct Meta { ... } // private struct shared between all operations, this could be adapted to hold/expose the entire HTTP response, e.g.

struct Foo { 
  /* regular fields */
  // private field
  _meta: Meta
}
```